sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -422,47 +422,487 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
   protected override def nullSafeEval(input: Any): Any = cast(input)
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    // TODO: Add support for more data types.
-    (child.dataType, dataType) match {
+    val eval = child.gen(ctx)
+    val nullSafeCast = nullSafeCastFunction(child.dataType, dataType, ctx)
+    if (nullSafeCast != null) {
+      eval.code +
+        castCode(ctx, eval.primitive, eval.isNull, ev.primitive, ev.isNull, dataType, nullSafeCast)
+    } else {
+      super.genCode(ctx, ev)
```

Comment:
`super.genCode(ctx, ev)` will call `child.gen(ctx)` again, you should move the `child.gen(ctx)` into `if` or `castCode`.
How about define `castCode` like `def castCode(ctx: CodeGenContext, ev: GeneratedExpressionCode, cast: CastFunction)`?